docs/mr_proposal.md

```diff
@@ -0,0 +1,187 @@
+# CUDA `<memory_resource>` Extension Proposal
+
+## Motivation
+
+Performance sensitive applications that make frequent dynamic memory allocations often find that allocating memory to be a significant overhead. 
+CUDA developers are even more acutely aware of the costs of dynamic allocation due to the relatively higher cost of `cudaMalloc/cudaFree` compared to standard `malloc/free`.
+As a result, developers devise custom, high-performance memory allocators as optimized as the application the allocator serves. 
+However, what works well for one application will not always satisfy another, which leads to a proliferation of custom allocator implementations. 
+Interoperation among these applications is difficult without an interface to enable sharing a common allocator.
+
+In Standard C++, [`Allocator`s](https://en.cppreference.com/w/cpp/named_req/Allocator) have traditionally provided this common interface.
+C++17 introduced [`<memory_resource>`](https://en.cppreference.com/w/cpp/header/memory_resource) and the [`std::pmr::memory_resource`](https://en.cppreference.com/w/cpp/memory/memory_resource) abstract class that defines a minimal interface for (de)allocating raw bytes and sits below `Allocator`s. 
+This polymorphic interface provides the lingua franca for those who trade in custom memory allocators. 
+
+
+<!--- In addition, `<memory_resource>` provides a handful of standard `memory_resource` implementations akin to custom allocators that seek to perform better than standard allocation, e.g., [`unsynchronized_pool_resource`](https://en.cppreference.com/w/cpp/memory/unsynchronized_pool_resource). --->
+
+However, the `std::pmr::memory_resource` interface is insufficient to capture the unique features of the CUDA C++ programming model.
+For example, Standard C++ only recognizes a single, universally accessible memory space; whereas CUDA C++ applications trade in at least four different kinds of dynamically allocated memory.
+Furthermore, CUDA's "stream"-based asynchronous execution model was extended in CUDA 11.2 with the addition of [`cudaMallocAsync`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html)<sup>[1](#link-footnote)</sup> to include memory allocation as stream-ordered events.
```

Comment:
Event might be overloaded. Maybe just introduce the term "stream-ordered memory allocation here".

```suggestion
Furthermore, CUDA's "stream"-based asynchronous execution model was extended in CUDA 11.2 with the addition of [`cudaMallocAsync` and `cudaFreeAsync`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html)<sup>[1](#link-footnote)</sup>, which provide stream-ordered memory allocation and deallocation.
```